src/main/scala/org/ergoplatform/nodeView/wallet/Registry.scala

```diff
@@ -0,0 +1,131 @@
+package org.ergoplatform.nodeView.wallet
+
+import io.iohk.iodb.ByteArrayWrapper
+import org.ergoplatform.nodeView.history.ErgoHistory.Height
+
+import scala.collection.mutable
+
+object Registry {
+
+  private val registry = mutable.Map[ByteArrayWrapper, TrackedBox]()
+
+  //todo: build indexes instead of iteration
+  def unspentBoxes: Iterator[UnspentBox] = registry.valuesIterator.collect {
+    case ub: UnspentBox => ub
+  }
+
+  def uncertainBoxes: Iterator[TrackedBox] = registry.valuesIterator.filterNot(_.certain)
+
+  //todo: extract a random element, not head
+  def nextUncertain(): Option[TrackedBox] = uncertainBoxes.toSeq.headOption
+
+  def registryContains(boxId: ByteArrayWrapper): Boolean = registry.contains(boxId)
+
+  def putToRegistry(trackedBox: TrackedBox) = synchronized {
+    registry.put(trackedBox.boxId, trackedBox)
+  }
+
+  def removeFromRegistry(boxId: ByteArrayWrapper) = registry.remove(boxId)
+
+  private val confirmedIndex = mutable.TreeMap[Height, Seq[ByteArrayWrapper]]()
+
+  def putToConfirmedIndex(height: Height, boxId: ByteArrayWrapper) = synchronized {
+    confirmedIndex.put(height, confirmedIndex.getOrElse(height, Seq.empty) :+ boxId)
+  }
+
+  def confirmedAt(height: Height): Seq[ByteArrayWrapper] = confirmedIndex.getOrElse(height, Seq.empty)
+
+  private var _confirmedBalance: Long = 0
+  private val _confirmedAssetBalances: mutable.Map[ByteArrayWrapper, Long] = mutable.Map()
+
+  private var _unconfirmedBalance: Long = 0
+  private val _unconfirmedAssetBalances: mutable.Map[ByteArrayWrapper, Long] = mutable.Map()
```

Comment:
we need a concurrent data structure here, or all reads should be synchronized as well as writes